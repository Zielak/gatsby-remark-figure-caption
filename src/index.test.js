require("object.entries/shim")(); // gatsby supports node v6
const Remark = require("remark");
const toHAST = require("mdast-util-to-hast");
const find = require("unist-util-find");
const hcheck = require("hast-util-assert");
const mcheck = require("mdast-util-assert");
const select = require("hast-util-select").selectAll;
const strip = require("unist-util-remove-position");
const render = require("unist-util-inspect");
const inputs = require("../fixtures/markdown");
const { name } = require("../package.json");
const plugin = require(".");
const remark = new Remark().data("settings", {
	commonmark: true,
	footnotes: true,
	pedantic: true,
});

const options = {
	figureClassName: "wrapper",
	captionClassName: "text",
	imageClassName: "lqip",
};

expect.addSnapshotSerializer({
	test: (tree) => tree && tree.type === "root",
	print: (tree) => render.noColor(tree),
});

const canon = (ast) => JSON.parse(JSON.stringify(strip(ast, true)));
const hasAlt = (node) => !!(node && node.properties && node.properties.alt);
const normal = (ast) => (typeof ast === "string" ? parse(ast) : ast);
const parse = (string) => canon(remark.parse(string));
const convert = (ast) =>
	canon(toHAST(normal(ast), { allowDangerousHTML: true }));
const transform = (markdownAST, options) =>
	typeof markdownAST === "string"
		? transform(normal(markdownAST), options)
		: (plugin({ markdownAST }, options), convert(markdownAST));

describe(name, () => {
	it("plugin exports a function", () => {
		expect(typeof plugin).toBe("function");
	});

	Object.entries(inputs).forEach(([description, string]) => {
		const before = convert(string);
		const result = transform(string);

		if (description === "WITHOUT_ALT") {
			it(`noop on images without alt [${description}]`, () => {
				expect(before).toMatchObject(result);
			});
		} else {
			it(`removes all alt properties [${description}]`, () => {
				expect(find(before, hasAlt)).toBeDefined();
				expect(find(result, hasAlt)).toBeUndefined();
			});
		}
	});

	![inputs.KITCHEN_SINK, inputs.PARAGRAPHS, inputs.INLINE].forEach(
		(markdown) => {
			const mdast_basic = normal(markdown);
			const mdast_class = normal(markdown);
			it("snapshot: mdast_basic", () => expect(mdast_basic).toMatchSnapshot());
			it("snapshot: mdast_class", () => expect(mdast_basic).toMatchSnapshot());
			it("must be valid markdown AST", () => {
				expect(() => mcheck(hast_basic)).not.toThrow();
				expect(() => mcheck(hast_class)).not.toThrow();
			});

			const hast_basic = transform(mdast_basic);
			const hast_class = transform(mdast_class, options);
			it("must be valid html AST", () => {
				expect(() => hcheck(hast_basic)).not.toThrow();
				expect(() => hcheck(hast_class)).not.toThrow();
			});
			it("snapshot: hast_basic", () => expect(hast_basic).toMatchSnapshot());
			it("snapshot: hast_class", () => expect(hast_class).toMatchSnapshot());
		}
	);

	test.each`
		key                    | count
		${"WITHOUT_ALT"}       | ${0}
		${"WITH_ALT"}          | ${1}
		${"INSIDE_BLOCKQUOTE"} | ${1}
		${"INLINE"}            | ${1}
		${"KITCHEN_SINK"}      | ${3}
		${"PARAGRAPHS"}        | ${3}
	`("$key => $count transformations", ({ key, count }) => {
		const mdast = normal(inputs[key]);
		const hast = transform(mdast, options);
		const { figureClassName, captionClassName, imageClassName } = options;
		expect(select(`figure.${figureClassName}`, hast)).toHaveLength(count);
		expect(select(`img.${imageClassName}`, hast)).toHaveLength(count);
		expect(select(`figcaption.${captionClassName}`, hast)).toHaveLength(count);
	});
});
