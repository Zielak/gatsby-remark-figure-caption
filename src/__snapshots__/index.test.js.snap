// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-remark-figure-caption snapshot: hast_basic 1`] = `
root[5]
├─ element[3] [tagName="p"]
│  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  ├─ text: "\\n"
│  └─ element[2] [tagName="figure"]
│     ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
│     └─ element[1] [tagName="figcaption"]
│        └─ text: "alt description"
├─ text: "\\n"
├─ element[3] [tagName="blockquote"]
│  ├─ text: "\\n"
│  ├─ element[9] [tagName="p"]
│  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  │  ├─ text: "\\n"
│  │  ├─ element[2] [tagName="figure"]
│  │  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
│  │  │  └─ element[1] [tagName="figcaption"]
│  │  │     └─ text: "alt description"
│  │  ├─ text: "\\n"
│  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  │  ├─ text: " "
│  │  ├─ element[2] [tagName="figure"]
│  │  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
│  │  │  └─ element[1] [tagName="figcaption"]
│  │  │     └─ text: "alt description"
│  │  ├─ text: "\\n"
│  │  └─ element[1] [tagName="code"]
│  │     └─ text: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
│  └─ text: "\\n"
├─ text: "\\n"
└─ element[1] [tagName="pre"]
   └─ element[1] [tagName="code"][properties={"className":["language-markdown"]}]
      └─ text: "![](https://example.com/image-without-alt.jpg)\\n![alt description](https://example.com/image.jpg)\\n> ![](https://example.com/image-without-alt.jpg)\\n> ![alt description](https://example.com/image.jpg)\\n"
`;

exports[`gatsby-remark-figure-caption snapshot: hast_basic 2`] = `
root[13]
├─ element[1] [tagName="p"]
│  └─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
├─ text: "\\n"
├─ element[1] [tagName="p"]
│  └─ element[2] [tagName="figure"]
│     ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
│     └─ element[1] [tagName="figcaption"]
│        └─ text: "alt description"
├─ text: "\\n"
├─ element[3] [tagName="blockquote"]
│  ├─ text: "\\n"
│  ├─ element[1] [tagName="p"]
│  │  └─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  └─ text: "\\n"
├─ text: "\\n"
├─ element[3] [tagName="blockquote"]
│  ├─ text: "\\n"
│  ├─ element[1] [tagName="p"]
│  │  └─ element[2] [tagName="figure"]
│  │     ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
│  │     └─ element[1] [tagName="figcaption"]
│  │        └─ text: "alt description"
│  └─ text: "\\n"
├─ text: "\\n"
├─ element[4] [tagName="p"]
│  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  ├─ text: " "
│  ├─ element[2] [tagName="figure"]
│  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
│  │  └─ element[1] [tagName="figcaption"]
│  │     └─ text: "alt description"
│  └─ text: " "
├─ text: "\\n"
├─ element[1] [tagName="p"]
│  └─ element[1] [tagName="code"]
│     └─ text: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
├─ text: "\\n"
└─ element[1] [tagName="pre"]
   └─ element[1] [tagName="code"][properties={"className":["language-markdown"]}]
      └─ text: "![](https://example.com/image-without-alt.jpg)\\n\\n![alt description](https://example.com/image.jpg)\\n\\n> ![](https://example.com/image-without-alt.jpg)\\n\\n> ![alt description](https://example.com/image.jpg)\\n"
`;

exports[`gatsby-remark-figure-caption snapshot: hast_basic 3`] = `
root[1]
└─ element[5] [tagName="p"]
   ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
   ├─ text: " "
   ├─ element[2] [tagName="figure"]
   │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg"}]
   │  └─ element[1] [tagName="figcaption"]
   │     └─ text: "alt description"
   ├─ text: "  "
   └─ element[1] [tagName="code"]
      └─ text: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
`;

exports[`gatsby-remark-figure-caption snapshot: hast_class 1`] = `
root[5]
├─ element[3] [tagName="p"]
│  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  ├─ text: "\\n"
│  └─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
│     ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
│     └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
│        └─ text: "alt description"
├─ text: "\\n"
├─ element[3] [tagName="blockquote"]
│  ├─ text: "\\n"
│  ├─ element[9] [tagName="p"]
│  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  │  ├─ text: "\\n"
│  │  ├─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
│  │  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
│  │  │  └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
│  │  │     └─ text: "alt description"
│  │  ├─ text: "\\n"
│  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  │  ├─ text: " "
│  │  ├─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
│  │  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
│  │  │  └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
│  │  │     └─ text: "alt description"
│  │  ├─ text: "\\n"
│  │  └─ element[1] [tagName="code"]
│  │     └─ text: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
│  └─ text: "\\n"
├─ text: "\\n"
└─ element[1] [tagName="pre"]
   └─ element[1] [tagName="code"][properties={"className":["language-markdown"]}]
      └─ text: "![](https://example.com/image-without-alt.jpg)\\n![alt description](https://example.com/image.jpg)\\n> ![](https://example.com/image-without-alt.jpg)\\n> ![alt description](https://example.com/image.jpg)\\n"
`;

exports[`gatsby-remark-figure-caption snapshot: hast_class 2`] = `
root[13]
├─ element[1] [tagName="p"]
│  └─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
├─ text: "\\n"
├─ element[1] [tagName="p"]
│  └─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
│     ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
│     └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
│        └─ text: "alt description"
├─ text: "\\n"
├─ element[3] [tagName="blockquote"]
│  ├─ text: "\\n"
│  ├─ element[1] [tagName="p"]
│  │  └─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  └─ text: "\\n"
├─ text: "\\n"
├─ element[3] [tagName="blockquote"]
│  ├─ text: "\\n"
│  ├─ element[1] [tagName="p"]
│  │  └─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
│  │     ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
│  │     └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
│  │        └─ text: "alt description"
│  └─ text: "\\n"
├─ text: "\\n"
├─ element[4] [tagName="p"]
│  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
│  ├─ text: " "
│  ├─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
│  │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
│  │  └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
│  │     └─ text: "alt description"
│  └─ text: " "
├─ text: "\\n"
├─ element[1] [tagName="p"]
│  └─ element[1] [tagName="code"]
│     └─ text: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
├─ text: "\\n"
└─ element[1] [tagName="pre"]
   └─ element[1] [tagName="code"][properties={"className":["language-markdown"]}]
      └─ text: "![](https://example.com/image-without-alt.jpg)\\n\\n![alt description](https://example.com/image.jpg)\\n\\n> ![](https://example.com/image-without-alt.jpg)\\n\\n> ![alt description](https://example.com/image.jpg)\\n"
`;

exports[`gatsby-remark-figure-caption snapshot: hast_class 3`] = `
root[1]
└─ element[5] [tagName="p"]
   ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image-without-alt.jpg"}]
   ├─ text: " "
   ├─ element[2] [tagName="figure"][properties={"className":["wrapper"]}]
   │  ├─ element[0] [tagName="img"][properties={"src":"https://example.com/image.jpg","className":["lqip"]}]
   │  └─ element[1] [tagName="figcaption"][properties={"className":["text"]}]
   │     └─ text: "alt description"
   ├─ text: "  "
   └─ element[1] [tagName="code"]
      └─ text: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
`;

exports[`gatsby-remark-figure-caption snapshot: mdast_basic 1`] = `
root[3]
├─ paragraph[3]
│  ├─ image [url="https://example.com/image-without-alt.jpg"]
│  ├─ text: "\\n"
│  └─ unknown[2] [data={"hName":"figure"}]
│     ├─ image [url="https://example.com/image.jpg"]
│     └─ paragraph[1] [data={"hName":"figcaption"}]
│        └─ text: "alt description"
├─ blockquote[1]
│  └─ paragraph[9]
│     ├─ image [url="https://example.com/image-without-alt.jpg"]
│     ├─ text: "\\n"
│     ├─ unknown[2] [data={"hName":"figure"}]
│     │  ├─ image [url="https://example.com/image.jpg"]
│     │  └─ paragraph[1] [data={"hName":"figcaption"}]
│     │     └─ text: "alt description"
│     ├─ text: "\\n"
│     ├─ image [url="https://example.com/image-without-alt.jpg"]
│     ├─ text: " "
│     ├─ unknown[2] [data={"hName":"figure"}]
│     │  ├─ image [url="https://example.com/image.jpg"]
│     │  └─ paragraph[1] [data={"hName":"figcaption"}]
│     │     └─ text: "alt description"
│     ├─ text: " \\n"
│     └─ inlineCode: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
└─ code: "![](https://example.com/image-without-alt.jpg)\\n![alt description](https://example.com/image.jpg)\\n> ![](https://example.com/image-without-alt.jpg)\\n> ![alt description](https://example.com/image.jpg)" [lang="markdown"]
`;

exports[`gatsby-remark-figure-caption snapshot: mdast_basic 2`] = `
root[7]
├─ paragraph[1]
│  └─ image [url="https://example.com/image-without-alt.jpg"]
├─ paragraph[1]
│  └─ unknown[2] [data={"hName":"figure"}]
│     ├─ image [url="https://example.com/image.jpg"]
│     └─ paragraph[1] [data={"hName":"figcaption"}]
│        └─ text: "alt description"
├─ blockquote[1]
│  └─ paragraph[1]
│     └─ image [url="https://example.com/image-without-alt.jpg"]
├─ blockquote[1]
│  └─ paragraph[1]
│     └─ unknown[2] [data={"hName":"figure"}]
│        ├─ image [url="https://example.com/image.jpg"]
│        └─ paragraph[1] [data={"hName":"figcaption"}]
│           └─ text: "alt description"
├─ paragraph[4]
│  ├─ image [url="https://example.com/image-without-alt.jpg"]
│  ├─ text: " "
│  ├─ unknown[2] [data={"hName":"figure"}]
│  │  ├─ image [url="https://example.com/image.jpg"]
│  │  └─ paragraph[1] [data={"hName":"figcaption"}]
│  │     └─ text: "alt description"
│  └─ text: " "
├─ paragraph[1]
│  └─ inlineCode: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
└─ code: "![](https://example.com/image-without-alt.jpg)\\n\\n![alt description](https://example.com/image.jpg)\\n\\n> ![](https://example.com/image-without-alt.jpg)\\n\\n> ![alt description](https://example.com/image.jpg)" [lang="markdown"]
`;

exports[`gatsby-remark-figure-caption snapshot: mdast_basic 3`] = `
root[1]
└─ paragraph[5]
   ├─ image [url="https://example.com/image-without-alt.jpg"]
   ├─ text: " "
   ├─ unknown[2] [data={"hName":"figure"}]
   │  ├─ image [url="https://example.com/image.jpg"]
   │  └─ paragraph[1] [data={"hName":"figcaption"}]
   │     └─ text: "alt description"
   ├─ text: "  "
   └─ inlineCode: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
`;

exports[`gatsby-remark-figure-caption snapshot: mdast_class 1`] = `
root[3]
├─ paragraph[3]
│  ├─ image [url="https://example.com/image-without-alt.jpg"]
│  ├─ text: "\\n"
│  └─ unknown[2] [data={"hName":"figure"}]
│     ├─ image [url="https://example.com/image.jpg"]
│     └─ paragraph[1] [data={"hName":"figcaption"}]
│        └─ text: "alt description"
├─ blockquote[1]
│  └─ paragraph[9]
│     ├─ image [url="https://example.com/image-without-alt.jpg"]
│     ├─ text: "\\n"
│     ├─ unknown[2] [data={"hName":"figure"}]
│     │  ├─ image [url="https://example.com/image.jpg"]
│     │  └─ paragraph[1] [data={"hName":"figcaption"}]
│     │     └─ text: "alt description"
│     ├─ text: "\\n"
│     ├─ image [url="https://example.com/image-without-alt.jpg"]
│     ├─ text: " "
│     ├─ unknown[2] [data={"hName":"figure"}]
│     │  ├─ image [url="https://example.com/image.jpg"]
│     │  └─ paragraph[1] [data={"hName":"figcaption"}]
│     │     └─ text: "alt description"
│     ├─ text: " \\n"
│     └─ inlineCode: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
└─ code: "![](https://example.com/image-without-alt.jpg)\\n![alt description](https://example.com/image.jpg)\\n> ![](https://example.com/image-without-alt.jpg)\\n> ![alt description](https://example.com/image.jpg)" [lang="markdown"]
`;

exports[`gatsby-remark-figure-caption snapshot: mdast_class 2`] = `
root[7]
├─ paragraph[1]
│  └─ image [url="https://example.com/image-without-alt.jpg"]
├─ paragraph[1]
│  └─ unknown[2] [data={"hName":"figure"}]
│     ├─ image [url="https://example.com/image.jpg"]
│     └─ paragraph[1] [data={"hName":"figcaption"}]
│        └─ text: "alt description"
├─ blockquote[1]
│  └─ paragraph[1]
│     └─ image [url="https://example.com/image-without-alt.jpg"]
├─ blockquote[1]
│  └─ paragraph[1]
│     └─ unknown[2] [data={"hName":"figure"}]
│        ├─ image [url="https://example.com/image.jpg"]
│        └─ paragraph[1] [data={"hName":"figcaption"}]
│           └─ text: "alt description"
├─ paragraph[4]
│  ├─ image [url="https://example.com/image-without-alt.jpg"]
│  ├─ text: " "
│  ├─ unknown[2] [data={"hName":"figure"}]
│  │  ├─ image [url="https://example.com/image.jpg"]
│  │  └─ paragraph[1] [data={"hName":"figcaption"}]
│  │     └─ text: "alt description"
│  └─ text: " "
├─ paragraph[1]
│  └─ inlineCode: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
└─ code: "![](https://example.com/image-without-alt.jpg)\\n\\n![alt description](https://example.com/image.jpg)\\n\\n> ![](https://example.com/image-without-alt.jpg)\\n\\n> ![alt description](https://example.com/image.jpg)" [lang="markdown"]
`;

exports[`gatsby-remark-figure-caption snapshot: mdast_class 3`] = `
root[1]
└─ paragraph[5]
   ├─ image [url="https://example.com/image-without-alt.jpg"]
   ├─ text: " "
   ├─ unknown[2] [data={"hName":"figure"}]
   │  ├─ image [url="https://example.com/image.jpg"]
   │  └─ paragraph[1] [data={"hName":"figcaption"}]
   │     └─ text: "alt description"
   ├─ text: "  "
   └─ inlineCode: "![](https://example.com/image-without-alt.jpg) ![alt description](https://example.com/image.jpg)"
`;
